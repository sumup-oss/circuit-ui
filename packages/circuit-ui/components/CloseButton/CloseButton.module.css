.base {
  /* The !important is necessary to override the default hover styles */
  border-color: transparent !important;
}
