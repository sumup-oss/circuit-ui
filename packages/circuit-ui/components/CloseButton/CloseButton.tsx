/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { forwardRef } from 'react';
import { Close } from '@sumup/icons';

import { clsx } from '../../styles/clsx.js';
import { IconButton, IconButtonProps } from '../IconButton/IconButton.js';

import classes from './CloseButton.module.css';

export type CloseButtonProps = Omit<IconButtonProps, 'children'>;

/**
 * A generic close button.
 */
export const CloseButton = forwardRef<any, CloseButtonProps>(
  ({ label = 'Close', className, ...props }, ref) => (
    <IconButton
      type="button"
      className={clsx(classes.base, className)}
      label={label}
      {...props}
      ref={ref}
    >
      <Close size="16" />
    </IconButton>
  ),
);

CloseButton.displayName = 'CloseButton';
