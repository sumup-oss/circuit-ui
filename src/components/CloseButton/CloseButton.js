/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { size } from 'polished';
import { Cross } from '@sumup/icons';

import IconButton from '../IconButton';

const StyledIconButton = styled(IconButton)(
  ({ theme }) => css`
    label: close-button;
    ${size(theme.iconSizes.mega)};
  `
);

/**
 * A generic close button.
 */
const CloseButton = props => (
  <StyledIconButton {...props}>
    <Cross />
  </StyledIconButton>
);

CloseButton.propTypes = {
  /**
   * Text label for screen readers. Important for accessibility.
   */
  label: PropTypes.string
};

CloseButton.defaultProps = {
  label: 'Close'
};

/**
 * @component
 */
export default CloseButton;
