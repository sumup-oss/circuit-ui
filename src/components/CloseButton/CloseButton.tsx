/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { css } from '@emotion/core';
import { Cross } from '@sumup/icons';

import { IconButton, IconButtonProps } from '../IconButton/IconButton';

export type CloseButtonProps = Omit<IconButtonProps, 'children'>;

const buttonStyles = () => css`
  border: 0;
`;

/**
 * A generic close button.
 */
export const CloseButton = React.forwardRef(
  (
    { label = 'Close', ...props }: CloseButtonProps,
    ref: CloseButtonProps['ref'],
  ) => (
    <IconButton
      type="button"
      css={buttonStyles}
      label={label}
      {...props}
      ref={ref}
    >
      <Cross />
    </IconButton>
  ),
);

CloseButton.displayName = 'CloseButton';
