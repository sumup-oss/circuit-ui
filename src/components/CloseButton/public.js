export { default as CloseButton } from '.';
