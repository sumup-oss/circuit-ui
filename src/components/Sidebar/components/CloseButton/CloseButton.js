/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import BaseButton from '../../../CloseButton';

const baseStyles = ({ theme }) => css`
  label: sidebar-close-button;
  border-radius: ${theme.borderRadius.circle};
  position: absolute;
  bottom: ${theme.spacings.mega};
  right: ${theme.spacings.mega};
  transition: opacity 150ms ease-in-out, visibility 150ms ease-in-out;
  visibility: hidden;
  opacity: 0;
  z-index: ${theme.zIndex.sidebar};
  ${theme.mq.giga} {
    visibility: hidden;
  }
`;

const visibleStyles = ({ visible }) =>
  visible &&
  css`
    label: close-button--visible;
    visibility: visible;
    opacity: 1;
  `;

const CloseButton = styled(BaseButton)`
  ${baseStyles};
  ${visibleStyles};
`;

CloseButton.propTypes = {
  /**
   * Tells if the CloseButton is visible
   */
  visible: PropTypes.bool
};

export default CloseButton;
