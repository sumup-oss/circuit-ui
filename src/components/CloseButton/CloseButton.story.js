import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import withTests from '../../util/withTests';
import CloseButton from './CloseButton';

storiesOf('CloseButton', module)
  .addDecorator(withTests('CloseButton'))
  .add('Default CloseButton', withInfo()(() => <CloseButton />));
