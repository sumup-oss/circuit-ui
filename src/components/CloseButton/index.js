import CloseButton from './CloseButton';

export default CloseButton;
