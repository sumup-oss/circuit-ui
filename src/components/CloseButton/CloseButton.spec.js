/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import CloseButton from '.';

describe('CloseButton', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<CloseButton />);
    expect(actual).toMatchSnapshot();
  });

  describe('business logic', () => {
    /**
     * Should accept a working ref
     */
    it('should accept a working ref', () => {
      const tref = React.createRef();
      const { container } = render(<CloseButton ref={tref} />);
      const button = container.querySelector('button');
      expect(tref.current).toBe(button);
    });
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<CloseButton />);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
