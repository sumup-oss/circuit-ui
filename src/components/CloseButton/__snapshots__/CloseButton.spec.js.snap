// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseButton should render with default styles 1`] = `
.circuit-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 120ms ease-in-out, color 120ms ease-in-out, background-color 120ms ease-in-out, border-color 120ms ease-in-out;
  transition: opacity 120ms ease-in-out, color 120ms ease-in-out, background-color 120ms ease-in-out, border-color 120ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
  border-radius: 8px;
  border: 0;
}

.circuit-1:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-1:focus::-moz-focus-inner {
  border: 0;
}

.circuit-1:disabled,
.circuit-1[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-1:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-1:active {
  background-color: #EEF0F2;
  border-color: #323E49;
}

.circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-1"
  title="Close"
>
  <svg
    fill="none"
    height="16"
    role="presentation"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 3l10 10m0-10L3 13"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-width="2"
    />
  </svg>
  <span
    class="circuit-0"
  >
    Close
  </span>
</button>
`;
